### PR TITLE
Implement YOLOF

### PR DESCRIPTION
**Pull request by hhaAndroid**
please refer to the paper of [You Only Look One-level Feature](https://arxiv.org/abs/2103.09460)

### CODE REVIEW

File: configs/yolof/yolof_r50_1x_coco.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+# optimizer
+optimizer = dict(
+    type='SGD',
+    lr=0.12,
+    momentum=0.9,
+    weight_decay=0.0001,
+    paramwise_cfg=dict(
+        norm_decay_mult=0.,
+        custom_keys={'backbone': dict(lr_mult=0.334, decay_mult=1.0)}))
+lr_config = dict(warmup_iters=1500, warmup_ratio=0.00066667)
```
Comment (ZwwWayne) [inline configs/yolof/yolof_r50_1x_coco.py:68]: We can directly use 1/3 and 2/3

File: configs/yolof/yolof_r50_1x_coco.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+            dict(type='Resize', keep_ratio=True),
+            dict(type='RandomFlip'),
+            dict(type='Normalize', **img_norm_cfg),
+            dict(type='Pad', size_divisor=32),
+            dict(type='ImageToTensor', keys=['img']),
+            dict(type='Collect', keys=['img']),
+        ])
+]
+data = dict(
+    samples_per_gpu=8,
```
Comment (ZwwWayne) [inline configs/yolof/yolof_r50_1x_coco.py:105]: If the default setting is not 8GPUSx2, we should rename the config, see convention in [doc](https://mmdetection.readthedocs.io/en/latest/tutorials/config.html)

File: mmdet/core/bbox/assigners/uniform_assigner.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
+import torch
+
+from ..builder import BBOX_ASSIGNERS
+from ..iou_calculators import build_iou_calculator
+from .assign_result import AssignResult
+from .base_assigner import BaseAssigner
+
+
+def box_xyxy_to_cxcywh(x):
```
Comment (ZwwWayne) [inline mmdet/core/bbox/assigners/uniform_assigner.py:9]: Check whether we already have a similar function

File: mmdet/datasets/pipelines/transforms.py [hunk 4]
```diff
@@ -472,6 +472,61 @@ def __repr__(self):
...
+    def __init__(self, prob=1.5, max_shift=32):
+        self.prob = prob
+        self.max_shift = max_shift
+
+    def __call__(self, results):
+        if random.random() < self.prob:
+            random_shift_x = random.randint(-self.max_shift, self.max_shift)
+            random_shift_y = random.randint(-self.max_shift, self.max_shift)
+            for key in results.get('img_fields', ['img']):
+                img = results[key]
```
Comment (ZwwWayne) [inline mmdet/datasets/pipelines/transforms.py:487]: May implement a new function to simplify the logic in the for loop.

File: mmdet/models/detectors/yolof.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
+from ..builder import DETECTORS
+from .single_stage import SingleStageDetector
+
+
+@DETECTORS.register_module()
+class YOLOF(SingleStageDetector):
+
```
Comment (ZwwWayne) [inline mmdet/models/detectors/yolof.py:7]: Add paper link

File: mmdet/models/necks/dilated_encoder.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+import torch.nn as nn
+from mmcv.cnn import caffe2_xavier_init
+from torch.nn import BatchNorm2d, ReLU
+
+from ..builder import NECKS
+
+
+class Bottleneck(nn.Module):
+
+    def __init__(self,
```
Comment (ZwwWayne) [inline mmdet/models/necks/dilated_encoder.py:11]: docstrings.

File: mmdet/models/necks/dilated_encoder.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+                 in_channels: int = 512,
+                 mid_channels: int = 128,
+                 dilation: int = 1):
+        super(Bottleneck, self).__init__()
+        self.conv1 = nn.Sequential(
+            nn.Conv2d(in_channels, mid_channels, kernel_size=1, padding=0),
```
Comment (ZwwWayne) [inline mmdet/models/necks/dilated_encoder.py:17]: Use ConvModule

File: mmdet/models/dense_heads/yolof_head.py [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+                    kernel_size=3,
+                    stride=1,
+                    padding=1))
+            bbox_subnet.append(BatchNorm2d(self.in_channels))
+            bbox_subnet.append(ReLU())
+        self.cls_subnet = nn.Sequential(*cls_subnet)
+        self.bbox_subnet = nn.Sequential(*bbox_subnet)
+        self.cls_score = nn.Conv2d(
+            self.in_channels,
+            self.num_anchors * self.num_classes,
```
Comment (ZephyrusKylin) [inline mmdet/models/dense_heads/yolof_head.py:102]: what if it is not sigmoid?

File: mmdet/core/bbox/assigners/uniform_assigner.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+from ..transforms import bbox_xyxy_to_cxcywh
+from .assign_result import AssignResult
+from .base_assigner import BaseAssigner
+
+
+@BBOX_ASSIGNERS.register_module()
+class UniformAssigner(BaseAssigner):
+
+    def __init__(self,
+                 pos_ignore_thresh,
```
Comment (ZwwWayne) [inline mmdet/core/bbox/assigners/uniform_assigner.py:14]: docstring

File: mmdet/core/bbox/assigners/uniform_assigner.py [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+        # Note that we use anchors and predict boxes both
+        cost_bbox = torch.cdist(
+            bbox_xyxy_to_cxcywh(bbox_pred),
+            bbox_xyxy_to_cxcywh(gt_bboxes),
+            p=1)
+        cost_bbox_anchors = torch.cdist(
+            bbox_xyxy_to_cxcywh(anchor), bbox_xyxy_to_cxcywh(gt_bboxes), p=1)
+
+        # TODO: TOPK function has different results in cpu and cuda mode
+        C = cost_bbox.cpu()
```
Comment (ZwwWayne) [inline mmdet/core/bbox/assigners/uniform_assigner.py:51]: Why we move the cost to CPU and conduct `topk` on CPU?

File: mmdet/models/necks/dilated_encoder.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
+import torch
+import torch.nn as nn
+from mmcv.cnn import caffe2_xavier_init
+from torch.nn import BatchNorm2d, ReLU
+
+from ..builder import NECKS
+
+
+class Bottleneck(nn.Module):
+    """Bottleneck block for DilatedEncoder.
```
Comment (ZwwWayne) [inline mmdet/models/necks/dilated_encoder.py:10]: May add paper reference.
```
Bottleneck block for DilatedEncoder used in `YOLOF <https://arxiv.org/abs/2103.09460>`_
```

File: mmdet/models/necks/dilated_encoder.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+        for m in [self.lateral_norm, self.fpn_norm]:
+            nn.init.constant_(m.weight, 1)
+            nn.init.constant_(m.bias, 0)
+        for m in self.dilated_encoder_blocks.modules():
+            if isinstance(m, nn.Conv2d):
+                nn.init.normal_(m.weight, mean=0, std=0.01)
+                if hasattr(m, 'bias') and m.bias is not None:
+                    nn.init.constant_(m.bias, 0)
+
+            if isinstance(m, (nn.GroupNorm, nn.BatchNorm2d, nn.SyncBatchNorm)):
```
Comment (ZwwWayne) [inline mmdet/models/necks/dilated_encoder.py:103]: we have is_norm in mmcv

File: mmdet/models/necks/dilated_encoder.py [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    The Bottleneck contains three ConvLayers and one residual connection.
+
+    Args:
+        in_channels (int): The number of input channels.
+        mid_channels (int): The number of middle output channels.
+        dilation (int): Dilation rate.
+    """
+
+    def __init__(self,
+                 in_channels: int = 512,
```
Comment (ZwwWayne) [inline mmdet/models/necks/dilated_encoder.py:21]: The number of convs should be settable.

File: mmdet/core/bbox/assigners/uniform_assigner.py (hunk 10 above)
Comment (hhaAndroid) [reply mmdet/core/bbox/assigners/uniform_assigner.py:51]: I refer to the open-source code, and then I found that the output results of CPU and GPU are different.

Comment (ZwwWayne) [reply mmdet/core/bbox/assigners/uniform_assigner.py:51]: If the difference does not affect the performance, we could use GPU for efficiency. No matter which choice, we should add comments to indicate the reason.

**Comment by ZwwWayne** [comment]
Resolve conflict

File: configs/yolof/iter_yolof_r50_c5_8x8_1x_coco.py [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+        norm_decay_mult=0.,
+        custom_keys={'backbone': dict(lr_mult=0.334, decay_mult=1.0)}))
+
+lr_config = dict(
+    policy='step',
+    warmup='linear',
+    warmup_iters=1500,
+    warmup_ratio=0.00066667,
+    step=[15000, 20000])
+runner = dict(_delete_=True, type='IterBasedRunner', max_iters=22500)
```
Comment (ZwwWayne) [inline configs/yolof/iter_yolof_r50_c5_8x8_1x_coco.py:76]: Add comment to indicate the real epoch number and how this number is calculated.

File: mmdet/datasets/pipelines/transforms.py [hunk 15]
```diff
@@ -472,6 +472,61 @@ def __repr__(self):
...
+                new_img = np.zeros_like(img)
+                if random_shift_x < 0:
```
Comment (ZwwWayne) [inline mmdet/datasets/pipelines/transforms.py:489]: May simplify the logic here.

File: configs/yolof/README.md [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+## Results and Models
+
+| Backbone  | Style   | Epoch | Lr schd | Mem (GB) |   box AP | Config | Download |
+|:---------:|:-------:|:-------:|:-------:|:--------:|:------:|:------:|:--------:|
+| R-50-C5     | caffe | Y | 1x      | 8.3      |   37.6  | [config](https://github.com/open-mmlab/mmdetection/tree/master/configs/yolof/yolof_r50_c5_8x8_1x_coco.py)       |[model]() &#124; [log]() |
+
+**Note**:
+
+1. We find that the performance is unstable and may fluctuate by about 0.3 mAP. mAP 37.4 ~ 37.7 is acceptable in YOLOF_R_50_C5_1x. Such fluctuation can also be found in the [original implementation](https://github.com/chensnathan/YOLOF).
+2. In addition to instability issues, I found that sometimes there are large loss fluctuations and NAN, so there may still be problems with this project, which will be improved subsequently.
```
Comment (ZwwWayne) [inline configs/yolof/README.md:25]: I found that sometimes there are large loss fluctuations and NAN -> sometimes there are large loss fluctuations and NAN,

File: configs/yolof/iter_yolof_r50_c5_8x8_1x_coco.py [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+
+# optimizer
+optimizer = dict(
+    type='SGD',
+    lr=0.12,
+    momentum=0.9,
+    weight_decay=0.0001,
+    paramwise_cfg=dict(
+        norm_decay_mult=0.,
+        custom_keys={'backbone': dict(lr_mult=1. / 3, decay_mult=1.0)}))
```
Comment (ZwwWayne) [inline configs/yolof/iter_yolof_r50_c5_8x8_1x_coco.py:70]: decay_mult do not need to be set if it is 1.0

File: configs/yolof/iter_yolof_r50_c5_8x8_1x_coco.py [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+
+# We implemented the iter-based config according to the source code.
+# COCO dataset has 117266 images after filtering. We use 8 gpu and
+# 8 batch size training, so 22500 is equivalent to
+# 22500/(117266/(8x8))=12.3 epoch, 15000 is equivalent to 8.2 epoch,
+# 20000 is equivalent to 10.9 epoch. Due to lr(0.12) is large,
+# the iter-based and epoch-based setting have about 0.2 influence on
```
Comment (ZwwWayne) [inline configs/yolof/iter_yolof_r50_c5_8x8_1x_coco.py:77]: influence -> difference

File: configs/yolof/iter_yolof_r50_c5_8x8_1x_coco.py [hunk 19]
```diff
@@ -0,0 +1,124 @@
+_base_ = [
```
Comment (ZwwWayne) [inline configs/yolof/iter_yolof_r50_c5_8x8_1x_coco.py:1]: 1. rename to yolof_r50_c5_8x8_iter-1x_coco.py
2. inherits from  yolof_r50_c5_8x8_1x_coco.py

File: mmdet/core/bbox/assigners/uniform_assigner.py [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+    Args:
+        pos_ignore_thresh (float): the threshold to ignore positive anchors
+        neg_ignore_thresh (float): the threshold to ignore negative anchors
+        match_times(int): Number of positive anchors for each gt box.
+           Default 4.
+        iou_calculator (dict): iou_calculator config
+    """
+
+    def __init__(self,
+                 pos_ignore_thresh,
```
Comment (ZwwWayne) [inline mmdet/core/bbox/assigners/uniform_assigner.py:26]: thresh -> thr as it is common in mmdet.

File: mmdet/core/bbox/assigners/uniform_assigner.py [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+        self.neg_ignore_thresh = neg_ignore_thresh
+        self.iou_calculator = build_iou_calculator(iou_calculator)
+
+    def assign(self,
+               bbox_pred,
+               anchor,
+               gt_bboxes,
+               gt_bboxes_ignore=None,
+               gt_labels=None):
+        warnings.warn('gt_bboxes_ignore was not considered for now.')
```
Comment (ZwwWayne) [inline mmdet/core/bbox/assigners/uniform_assigner.py:41]: gt_bboxes_ignore is always not considered. So this warning is unnecessary and can be put into the docstirng's Note section

File: mmdet/core/bbox/assigners/uniform_assigner.py [hunk 22]
```diff
@@ -0,0 +1,137 @@
...
+
+        # 4. Compute the ignore indexes of positive sample use anchors
+        # and predict boxes
+        pos_gt_index = torch.arange(
+            0, C1.size(1),
+            device=bbox_pred.device).repeat(self.match_times * 2)
+        pos_ious = anchor_overlaps[indexes, pos_gt_index]
+        pos_ignore_idx = pos_ious < self.pos_ignore_thresh
+
+        target_classes_o = pos_gt_index + 1
```
Comment (ZwwWayne) [inline mmdet/core/bbox/assigners/uniform_assigner.py:113]: target_classes_o stands for what? Add comments above it.

File: mmdet/models/dense_heads/yolof_head.py [hunk 23]
```diff
@@ -0,0 +1,414 @@
...
+                 num_classes,
+                 in_channels,
+                 cls_num_convs=2,
+                 reg_num_convs=4,
+                 norm_cfg=dict(type='BN', requires_grad=True),
+                 **kwargs):
+        self.cls_num_convs = cls_num_convs
+        self.reg_num_convs = reg_num_convs
+        self.norm_cfg = norm_cfg
+        self.INF = 1e8
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/yolof_head.py:63]: can directly put INF to the header like EPS in other files.

File: mmdet/models/dense_heads/yolof_head.py [hunk 24]
```diff
@@ -0,0 +1,414 @@
...
+           Default 2.
+        reg_num_convs (int): The number of convolutions of reg branch.
+           Default 4.
+        norm_cfg (dict): Dictionary to construct and config norm layer.
+    """
+
+    def __init__(self,
+                 num_classes,
+                 in_channels,
+                 cls_num_convs=2,
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/yolof_head.py:56]: num_cls_convs and num_reg_convs

File: mmdet/models/dense_heads/yolof_head.py [hunk 25]
```diff
@@ -0,0 +1,414 @@
...
+
+        flatten_labels = batch_labels.reshape(-1)
+        batch_label_weights = batch_label_weights.reshape(-1)
+        cls_score = cls_scores[0].permute(0, 2, 3,
+                                          1).reshape(-1, self.cls_out_channels)
+
+        num_total_samples = (num_total_pos +
+                             num_total_neg) if self.sampling else num_total_pos
+        num_total_samples = max(
+            reduce_mean(cls_score.new_tensor(num_total_samples)).item(), 1.0)
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/yolof_head.py:200]: -> reduce_mean(cls_score.new_tensor(num_total_samples)).clamp_(1.0).item()

File: mmdet/models/necks/dilated_encoder.py [hunk 26]
```diff
@@ -0,0 +1,107 @@
...
+from mmcv.cnn import (ConvModule, caffe2_xavier_init, constant_init, is_norm,
+                      normal_init)
+from torch.nn import BatchNorm2d
+
+from ..builder import NECKS
+
+
+class Bottleneck(nn.Module):
+    """Bottleneck block for DilatedEncoder used in `YOLOF.
+
```
Comment (ZwwWayne) [inline mmdet/models/necks/dilated_encoder.py:11]: Do not need empty line here, and '.' should be removed.

**Review by ZwwWayne — APPROVED** [review]
